File: Baekjoon/DP/2156.py
# https://www.acmicpc.net/problem/2156

import sys
input = sys.stdin.readline

n = int(input())
wines = [int(input()) for _ in range(n)]
answer = [0 for _ in range(n)]

if n >= 1:
    answer[0] = wines[0]
if n >= 2:
    answer[1] = wines[0] + wines[1]
if n >= 3:
    answer[2] = max(wines[0] + wines[1], wines[0] + wines[2], wines[1] + wines[2])
if n > 3:
    for i in range(3, n):
        answer[i] = max(answer[i - 1], answer[i - 2] + wines[i], answer[i - 3] + wines[i - 1] + wines[i])

print(answer[-1])

File: Baekjoon/DP/2156.cpp
// https://www.acmicpc.net/problem/2156

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int n;
    cin >> n;

    vector<int> wines(n);
    for (int i = 0; i < n; i++) {
        cin >> wines[i];
    }

    vector<vector<int>> dp(n, vector<int>(3, 0));
    dp[0][1] = wines[0];

    for (int i = 1; i < n; i++) {
        dp[i][0] = max({dp[i - 1][0], dp[i - 1][1], dp[i - 1][2]});
        dp[i][1] = dp[i - 1][0] + wines[i];
        dp[i][2] = dp[i - 1][1] + wines[i];
    }

    cout << max({dp[n - 1][0], dp[n - 1][1], dp[n - 1][2]}) << endl;

    return 0;
}